parsl/executors/high_throughput/mpi_resource_management.py

```diff
@@ -0,0 +1,222 @@
+        self.nodes_q: queue.Queue = queue.Queue()
+        for node in self.available_nodes:
+            self.nodes_q.put(node)
+
+        logger.info(
+            f"Starting MPITaskScheduler with {len(self.available_nodes)}"
+        )
+
+    def _get_nodes(self, num_nodes: int) -> List[str]:
+        """Thread safe method to acquire num_nodes from free resources
+
+        Raises: MPIResourceUnavailable if there aren't enough resources
+        Returns: List of nodenames:str
+        """
+        logger.debug(
+            f"Requesting : {num_nodes=} we have {self._free_node_counter}"
+        )
+        acquired_nodes = []
+        with self._free_node_counter.get_lock():
+            if num_nodes <= self._free_node_counter.value:  # type: ignore[attr-defined]
+                self._free_node_counter.value -= num_nodes  # type: ignore[attr-defined]
+            else:
+                raise MPIResourceUnavailable(
+                    requested=num_nodes, available=self._free_node_counter.value  # type: ignore[attr-defined]
+                )
+
+            for i in range(num_nodes):
+                node = self.nodes_q.get()
+                acquired_nodes.append(node)
+        return acquired_nodes
+
+    def _return_nodes(self, nodes: List[str]) -> None:
+        """Threadsafe method to return a list of nodes"""
+        for node in nodes:
+            self.nodes_q.put(node)
+        with self._free_node_counter.get_lock():
+            self._free_node_counter.value += len(nodes)  # type: ignore[attr-defined]
+
+    def put_task(self, task_package: dict):
+        """Schedule task if resources are available otherwise backlog the task"""
+        user_ns = locals()
+        user_ns.update({"__builtins__": __builtins__})
+        _f, _args, _kwargs, resource_spec = unpack_res_spec_apply_message(
+            task_package["buffer"], user_ns, copy=False
+        )
+
+        nodes_needed = resource_spec.get("NUM_NODES")
+        if nodes_needed:
+            try:
+                allocated_nodes = self._get_nodes(nodes_needed)
+            except MPIResourceUnavailable:
+                logger.warning("Not enough resources, placing task into backlog")
+                self._backlog_queue.put((nodes_needed, task_package))
+                return
+            else:
+                resource_spec["MPI_NODELIST"] = ",".join(allocated_nodes)
+                self._map_tasks_to_nodes[task_package["task_id"]] = allocated_nodes
+                buffer = pack_res_spec_apply_message(_f, _args, _kwargs, resource_spec)
+                task_package["buffer"] = buffer
+
+        self.pending_task_q.put(task_package)
+
+    def _schedule_backlog_tasks(self):
+        """Attempt to schedule backlogged tasks"""
+        try:
+            _nodes_requested, task_package = self._backlog_queue.get(block=False)
+            self.put_task(task_package)
+        except queue.Empty:
+            return
+        except MPIResourceUnavailable:
```

Comment:
does this `except` need to happen here? does self.put_task already handle rescheduling the task into the backlog again?

Reply:
Yes, the logic goes like this:
```
1. If the queue is empty -> return
```

Reply:
I'm suggesting that self.put_task will never raise an MPIResourceUnavailable exception so this except block will never fire.

Alternatively, if self.put_task can raise an MPIResourceUnavailable exception, then this exception should be protected against in the invocation `self.task_scheduler.put_task` in `Manager.pull_tasks` otherwise it will cause the Manager.pull_tasks loop to exit when resources are unavailable.

Reply:
(delete this dead code exception block)